### PR TITLE
providers/aws: Add support for policy on S3 bucket

### PR DESCRIPTION
**Pull request by justincampbell**
Adds support for policies on S3 buckets: http://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html

```c
resource "aws_s3_bucket" "default" {
  bucket = "hc-jc-tf-test"
  policy = "${file("policy.json")}"
}
```

### CODE REVIEW

File: builtin/providers/aws/resource_aws_s3_bucket_test.go [hunk 1]
```diff
@@ -145,6 +172,38 @@ func testAccCheckAWSS3BucketExists(n string) resource.TestCheckFunc {
...
+				// this error is expected
+				return nil
+			} else {
+				return fmt.Errorf("GetBucketPolicy error: %v", err)
+			}
+		}
+
+		if v := out.Policy; v == nil {
+			if policy != "" {
+				return fmt.Errorf("bad policy, found nil, expected: %s", policy)
```
Comment (justincampbell) [inline builtin/providers/aws/resource_aws_s3_bucket_test.go]: This currently fails due to https://github.com/awslabs/aws-sdk-go/issues/236

I believe @m-s-austin also ran into this in https://github.com/iJoinSolutions/terraform/commit/cd45e45d6d04a3956fe35c178d5e816ba18d1051

Comment (m-s-austin) [inline builtin/providers/aws/resource_aws_s3_bucket_test.go]: That does indeed appear to be bugged in the sdk, I had not had a chance to test it directly yet.

Comment (justincampbell) [inline builtin/providers/aws/resource_aws_s3_bucket_test.go]: This has been resolved in https://github.com/awslabs/aws-sdk-go/commit/c37b3cb43ea0ca2ee05432f620c06647491ba1dd

**Comment by justincampbell** [comment]
```
--- PASS: TestAccAWSS3Bucket_basic (3.80s)
--- PASS: TestAccAWSS3Bucket_Policy (4.22s)
--- PASS: TestAccAWSS3Bucket_Website (7.14s)
--- PASS: TestAccAWSS3Bucket_WebsiteRedirect (4.78s)
```

**Comment by m-s-austin** [comment]
@justincampbell I have been testing your code today, and one thing I noticed, in the version I was working on with a separate resource for the policy, I could delete the policy resource from the tf file, and then terraform apply would also delete the policy from the bucket. It seems that in this version I can not achieve the same functionality? Deleting the argument has no action and if I try to explicitly set policy = "" I get an error   policy: "" => "Error parsing JSON: unexpected end of JSON input" 

leaving me with no way to remove a policy from tf.

**Comment by catsby** [comment]
`TestAccAWSS3Bucket_Policy` is not passing for me:

```
provider.aws
--- FAIL: TestAccAWSS3Bucket_Policy (3.11s)
	testing.go:131: Step 0 error: Check failed: bad policy, found nil, expected: { "Version": "2008-10-17", "Statement": [ { "Sid": "", "Effect": "Allow", "Principal": { "AWS": "*" }, "Action": "s3:GetObject", "Resource": "arn:aws:s3:::tf-test-bucket-7742106432384881169/*" } ] }
FAIL
exit status 1
FAIL	github.com/hashicorp/terraform/builtin/providers/aws	3.126s
```

**Comment by justincampbell** [comment]
@m-s-austin Are you using the latest code from this PR? It seems to remove properly for me.

The `"" => "Error parsing JSON: unexpected end of JSON input"` behavior is unfortunate with invalid JSON. We're using `normalizeJson` as a `SchemaStateFunc`, which only returns a string: https://github.com/hashicorp/terraform/blob/master/helper/schema/schema.go#L174

I'm not sure what to do here. We could change `SchemaStateFunc` to also return an error. Or, just return an empty string, or possibly the original string passed in, on invalid JSON.

**Comment by justincampbell** [comment]
@catsby You may need to update aws-sdk-go

**Comment by catsby** [comment]
@justincampbell that did help, thanks :smile: 

It seems though that I have invalid JSON:

```
+ aws_s3_bucket.bucket
    acl:              "" => "public-read"
    bucket:           "" => "tf-test-bucket-14"
    hosted_zone_id:   "" => "<computed>"
    policy:           "" => "Error parsing JSON: invalid character '\\'' looking for beginning of object key string"
    region:           "" => "<computed>"
    website_endpoint: "" => "<computed>"
```

I do not want that policy applied :smile_cat:

**Comment by catsby** [comment]
That issue aside, I used a valid policy JSON doc (from a file) and it seemed to apply / remove as expected. 

@justincampbell did you test this with in-line policies? That's where I got the parsing error, I wonder if Go or HCL did something there. If you anticipate most people using from a file, we should provide show the  `"${file("./policy.json")}"` example in the docs

**Comment by m-s-austin** [comment]
@justincampbell I am using your latest, but I realized now I am not using aws-sdk-go with your bug fix, this would cause terraform to always refresh state to a null policy right?

**Comment by justincampbell** [comment]
@catsby Does a heredoc work? Otherwise you would have to escape all of the quotes

**Comment by m-s-austin** [comment]
I have tested with this format it works fine:
```
    policy = <<EOF
{
  "Id": "Policy1431447664360",
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Stmt1431447659874",
      "Action": [
        "s3:GetObject"
      ],
      "Effect": "Allow",
      "Resource": "arn:aws:s3:::my_tf_test_bucket/*",
      "Principal": "*"
    }
  ]
}
EOF
```

**Comment by justincampbell** [comment]
@m-s-austin Thanks for testing!

**Comment by justincampbell** [comment]
@catsby Added `policy = "#{file("policy.json")}"` to the website example

**Comment by catsby** [comment]
LGTM :+1:

**Pull request merged**